### ISSUE TITLE
First steps to implement an ENFORCE statement. 

### ISSUE THREAD

**Issue by hanjoosten** [AmpersandTarski/Ampersand#1204]
In issue #311 we have been discussing the ENFORCE statement. Until now, nothing of this has been implemented. However, I see the possibility to do some first steps, to harvest some low-hanging fruit. 
The inspiration is from the observation that when we use the exec engine, we have a few patterns that seem to occur over and over again. The pattern I see is when we want some relation to being populated automatically by the same contents as some expression. In such a case, two rules are written. One with an `InsPair` command, the other using the `DelPair` command. 
Writing these statements currently involves quite some boilerplate, and it is error-prone. The type checker isn't involved, so this can easily lead to nasty-to-diagnose runtime issues. 

An example of this pattern:

~~~.ampersand
ROLE ExecEngine MAINTAINS "Compute transitive closure of isa by Ins", "Compute transitive closure of isa by Del"

RULE "Compute transitive closure of isa by Ins" : isaPlus\/I |- isaStar
VIOLATION (TXT "{EX} InsPair;isaStar;Concept;", SRC I, TXT ";Concept;", TGT I)

RULE "Compute transitive closure of isa by Del" : isaStar |- isaPlus\/I
VIOLATION (TXT "{EX} DelPair;isaStar;Concept;", SRC I, TXT ";Concept;", TGT I)
~~~

The idea here is that the population of the relation `isaStar[Concept*Concept]` should automatically be kept equal to the population of the expression `isaPlus \/ I`

I would like to add some syntax to the ampersand language, so this can be expressed by the modeller. However, adding syntax is easy, but removing syntax after a while is close to impossible. Therefore, we need to think about this syntax so we can be sure that we are not going to regret it. Having the ENFORCE statement issue in mind, I am now thinking to have it like this: 

~~~.ebnf
EnforceStatement
    ::= ENFORCE ( '<rulename>' ( ',' '<rulename>' )* ( INVARIANT | BY '<rolename>' ( ',' '<rolename>' )* ) 
                | COMPUTING '<relation>' ( ':=' | ':<' | '>:' ) '<expression>' 
                )
~~~
Which can be [visualized](https://bottlecaps.de/rr/ui) as
![image](https://user-images.githubusercontent.com/1159672/130403852-70d966af-5be8-4977-b98a-332db0cf2cd7.png)

I intend to start with implementing the `COMPUTING` part. 

A note on the operators: `:=` is used as an assignment in many computer languages. It seems a logical choice, as @WolframKahl also suggested in the earlier discussion. If that operator is used, both the `InsPair` and `DelPair` actions will be in place. `:<` will only trigger `DelPair`.  `>:` will only trigger `InsPair` .

**Comment by hanjoosten** [comment]
@stefjoosten , @RieksJ , @Michiel-s & @sjcjoosten , your thoughts please

**Comment by RieksJ** [comment]
Let me start to say that I like the idea of this, and @hanjoosten 's setup is a good start. I have some comments of different nature that I'll document in subsequent comments, which I'll label as RJ1, RJ2 etc. so that they may be easily referenced.

**Comment by RieksJ** [comment]
RJ1: Looking at source code that I have, I seen many instances where not one, but multiple relations are populated.

Suggestion: Allow `COMPUTING` to be followed by a list of `<relation>` `<operator>` `expression`.

**Comment by RieksJ** [comment]
RJ2: Let `COMPUTING` also cater for different kinds of processing, e.g. in some programming language. I would think @Michiel-s  would have some useful insights on this.

**Comment by RieksJ** [comment]
RJ3: make `INVARIANT` the default (and optional), meaning that if it isn't followed by `BY` or `COMPUTING`, this implies `INVARIANT`.

**Comment by RieksJ** [comment]
RJ4: It just occurred to me that the syntax `ENFORCE COMPUTING r := s;t` is allowed, and then there is no (need for a) `<rulename>`. This suggests that `COMPUTING` could be a statement in its own right. 

The proposed syntax is a mix of syntax that associates attributes (e.g. a role, or an 'INVARIANT' tag) to a real/named rule, and a COMPUTING statement that isn't about rules (at least not in the same way as as the part that associated attributes)

I suggest that we either
- split the syntax so that `COMPUTING` becomes a thing in its own, or
- completely merge the syntaxes for defining rules (and their attributes) and COMPUTING, which was what I had in mind in when drafting [RJ1](https://github.com/AmpersandTarski/Ampersand/issues/1204#issuecomment-902445222)

**Comment by RieksJ** [comment]
RJ5: `ROLE`s identify a class of (digital or human) actors that are tasked by the enforcement of the rule(s). It seems natural to provide each such class with some texts that they can interpret and use as a guide for performing this task. For roles that are assigned to human actors, this could take the form of a working instruction. For roles that are assigned to digital actors, this could take the form of executable instructions (depending on what such actors can do).

I'm not saying that we should immediately support all sorts of weird things, but thinking about the idea and making sure we can extend in these directions seems reasonable

**Comment by hanjoosten** [comment]
Take in mind that the purpose of this issue is to be able to make some first steps with the ENFORCE statement, and at the same time be aware of introducing syntax that we will regret in the future. That said, referring to the comments of @RieksJ :

- [RJ1:](https://github.com/AmpersandTarski/Ampersand/issues/1204#issuecomment-902445222) This could be done quite easily, however it will make it harder to refer to the exact location in case of (type)errors. So initially, I suggest not to support it.
- [RJ3:](https://github.com/AmpersandTarski/Ampersand/issues/1204#issuecomment-902446706) I see no use for that. If you want to explicitly write down that you want to enforce a rule, then this is how to do it. If you don't want to write it down explicitly, then skip the ENFORCE statement for that rule alltogether. 
- [RJ4:](https://github.com/AmpersandTarski/Ampersand/issues/1204#issuecomment-902449291) When the user writes down `ENFORCE COMPUTING r := s;t`, under the hood two rules will generated automatically. Splitting the syntax will make this issue more complex. However, in this case we could easily choose to remove the `COMPUTING` keyword, or to make that optional. Technically, the `COMPUTING` keyword has no use in the proposed syntax.
- [RJ5:](https://github.com/AmpersandTarski/Ampersand/issues/1204#issuecomment-902451162) When time comes (not this issue) we might want to allow the user to specify such things as properties of a `ROLE`. That will probably require syntax extentions too, that could be done with some `ROLE` statement. The `BY` part here only refers to a `ROLE`, so we can easily leave this out of scope for now, without jeopardizing such posibilities in the future.

**Comment by RieksJ** [comment]
Re @hanjoosten 's comment on RJ5: The current `VIOLATION` statement can already be seen as providing such guidance (in a limited way).

**Comment by sjcjoosten** [comment]
I'm confused about the keyword 'ENFORCE' here. I thought that the ENFORCE keyword ties a rule to how it should be enforced, which means that anything that starts with ENFORCE should not introduce new rules under the hood. Anything starting with ENFORCE should also not trigger any actions unless the rule is violated, and the purpose of the ENFORCE action is to (eventually) ensure that the rule is no longer violated.

I do recognize that quite often, the rule will be equal to the thing being enforced, and we might want special syntax for that, but I suggest distinguishing keywords that add things to rules (like MEANING and ENFORCE) to keywords that introduce new rules by themselves (like RULE, and perhaps something new like COMPUTED, MACRO or ABBREVIATION).

Note that we can use ENFORCE with a single rule too, if we could only:
- generate multiple violations
- tune the violations we see per violation in a style similar to what happens in interfaces. Here are my thoughts in pseudo-syntax:

```
RULE "Compute transitive closure of isa" : isaStar = isaPlus\/I
VIOLATION (I /\ (SRC;isaStar;TGT~)) BOX(TXT "{EX} DelPair;isaStar;Concept;", SRC I, TXT ";Concept;", TGT I) 
VIOLATION (I - (SRC;isaStar;TGT~)) BOX(TXT "{EX} InsPair;isaStar;Concept;", SRC I, TXT ";Concept;", TGT I) 
```

To understand the above, I have the following meta-model in mind:

```
SRC :: Violation -> Concept
TGT :: Violation -> Concept
```

I'm not proposing we build any of this now, just trying to show that we can ENFORCE on just a single rule. I would propose this syntax instead:

```
EnforceStatement
    ::= ENFORCE ( '<rulename>' ( ',' '<rulename>' )* ( INVARIANT | BY '<rolename>' ( ',' '<rolename>' )* 
                | COMPUTING '<relation>' ( ':=' | ':<' | ':>' ) '<expression>' )
                )
```

Another thing to note, is that we don't necessarily need an ENFORCE statement to fix all errors: the highly structured nature of 'r := e' makes it relatively easy to reason about whether we've exhausted all possibly computation.

**Comment by hanjoosten** [comment]
Above EBNF expression can be visualized as:
![image](https://user-images.githubusercontent.com/1159672/130354056-cdce8903-4f81-4954-92b8-4ca835f95ac7.png)

**Comment by hanjoosten** [comment]
I made this issue because we are debating about syntax, while that is the exact cause why there has not been done any work on this for a way too long time. 

The point is that the rules we want to automate have no name. I agree with @sjcjoosten that therefor the ENFORCE keyword looks odd. There are other places where we define rules implicitly. This is done for all `PropertyRules`. Under the hood, these rules have names too, which are generated. 

The issue we want (and should!) implement is about populating a single named relation. In the previous examples we want to be able to:

1)  automatically delete pairs from the population of a named relaltion if they are not in a given expression
2) automatically insert pairs into the population of a named relation if they are in a given expression
3) combine both above rules in a single statement.

There is no need at all to have the modeller specify a name for such rules. And even if we would opt to do so, the `:=` would require two names, because it is a combination of the `:<` and `:>` rules. Therefor the `COMPUTING` part should not have a mandatory `rulename`.

**Comment by hanjoosten** [comment]
So where does this all lead to? 

* The functionality to implement these rules is allready in place using the execengine. 
* The current syntax is very combersome and errorprone, as I illustrated above with the examples. 
* We have done nothing for over 5 years because we cannot agree about the syntax.
* Technically, the syntax doesn't matter a lot
* For users, syntax matters a lot. Once it is in place, it causes pain to remove it. A lot of computer languages suffer from this.
* There is quite some "business value" in being able to get this working. 

I would urge @stefjoosten to decide on what to do. 

# Scenario's

## 1) Implement the `COMPUTING` part 
This could easily be done as 

~~~
SomeStatement
         ::= SOMEKEYWORD '<relationname>' ( ':=' | ':<' | ':>' ) '<expression>'
~~~
Which can be [visualized](https://bottlecaps.de/rr/ui) as
![image](https://user-images.githubusercontent.com/1159672/130355687-bdd18848-77d8-4dd5-a892-172576e97a30.png)


The keyword can be decided upon while implementing. Note however, that `ENFORCE` can be a very logical choice: This statement denotes in a declarative way some enforcement of the population of the named relation.

If we choose `ENFORCE`  for the keyword, this leaves room to extend the statement later on. We could later go for something like:

~~~
EnforceStatement
         ::= ENFORCE ( '<relationname>' ( ':=' | ':<' | ':>' ) '<expression>' 
                     | '<rulename>' ( ',' '<rulename>' )* ( INVARIANT 
                                                          | BY '<rolename>' ( ',' '<rolename>' )*
                                                          )
                     )
~~~
Which can be [visualized](https://bottlecaps.de/rr/ui) as
![image](https://user-images.githubusercontent.com/1159672/130355978-7a16c347-d65f-4810-9656-2562a868b447.png)

 * Pros: This is low hanging fruit, with quite some value
 * Cons: We might regret the syntax?

## 2) Discuss forever.
 * Pros: We might eventually see the holy grail
 * Cons: We probably wait forever.

## 3) Some other scenario
See scenario 2).

**Comment by sjcjoosten** [comment]
> I made this issue because we are debating about syntax, while that is the exact cause why there has not been done any work on this for a way too long time.

Sorry, my bad. I thought it would be helpful and appreciated to get this input. I didn't realize that a lack of agreement over syntax was perceived of as a reason for why nothing was done on this, I never perceived it as such.

**Comment by hanjoosten** [comment]
No worries. I really appreciate your input on this. No doubt about it. On the other hand it is my frustration that an undecided discussion about syntax causes no progression.

**Comment by stefjoosten** [comment]
I Agree with scenario 1. Let's proceed along this path.

**Comment by RieksJ** [comment]
@WolframKahl made some [notes](https://github.com/AmpersandTarski/Ampersand/issues/311#issuecomment-184472195) on this that we should at least think about a bit, e.g. what happens if `<expression>` contains `<relationname>`?

**Comment by RieksJ** [comment]
I wouldn't mind forbidding `<relationname>` to occur in an `<expression>` in the syntax `<relationname> := <expression>`.

**Comment by stefjoosten** [comment]
> @WolframKahl made some [notes](https://github.com/AmpersandTarski/Ampersand/issues/311#issuecomment-184472195) on this that we should at least think about a bit, e.g. what happens if `<expression>` contains `<relationname>`?

That is recursion. The principle is well understood. For this purpose, the ExecEngine has a limit on the number of runs, so infinite recursion is blocked. Recursion is already possible in the current setup, so we do not have a new situation here. The new situation specializes the existing situation, and that might of course lead us to additional insights.

**Comment by hanjoosten** [comment]
I thought of that. This can also occur on a deeper way: `r := s;t`, `s:= t \/ r` etc. Forbidding could be a way to go, but initially I think warnings would be appropriate.

**Comment by stefjoosten** [comment]
A compile-time warning will do. The effect of a limit on the number of ExecEngine runs will be that certain violations are not resolved once the ExecEngine quits. That is quite acceptable behavior, especially if it is accompanied by warnings. (To let the ExecEngine recur infinitely would be acceptable only if the end-user has a way to interrupt the ExecEngine, which is not the case currently.) To discuss the details wrt recursion is maybe better done in a separate issue.

**Comment by DebbieGit** [comment]
Interessante discussie,
Debbie

**Comment by hanjoosten** [comment]
Closed with PR https://github.com/AmpersandTarski/Ampersand/pull/1207

**Comment by stefjoosten** [comment]
@hanjoosten Where did you document the final syntax that you have implemented?

**Comment by hanjoosten** [comment]
> @hanjoosten Where did you document the final syntax that you have implemented?

https://app.gitbook.com/@ampersandtarski/s/documentation/~/drafts/-Mj14B-i9i5vadHKSYVq/the-language-ampersand/the-enforce-statement
